sys/aarch64/pl011.c

```diff
@@ -0,0 +1,244 @@
+};
+/* clang-format on */
+
+static intr_filter_t pl011_intr(void *data /* device_t* */) {
+  pl011_state_t *state = data;
+  intr_filter_t res = IF_STRAY;
+
+  WITH_SPIN_LOCK (&state->lock) {
+    if (pl011_rready(state)) {
+      ringbuf_putb(&state->rx_buf, pl011_getc(state));
+      cv_signal(&state->rx_nonempty);
+      res = IF_FILTERED;
+    }
+
+    if (pl011_wready(state)) {
+      uint8_t byte;
+      if (ringbuf_getb(&state->tx_buf, &byte)) {
+        pl011_putc(state, byte);
+        cv_signal(&state->tx_nonfull);
+      }
+      res = IF_FILTERED;
+    }
+  }
+
+  return res;
+}
+
+static int pl011_probe(device_t *dev) {
+  /* (pj) so far we don't have better way to associate driver with device for
+   * buses which do not automatically enumerate their children. */
+  return (dev->unit == 1);
+}
+
+static int pl011_attach(device_t *dev) {
+  pl011_state_t *state = dev->state;
+
+  ringbuf_init(&state->rx_buf, kmalloc(M_DEV, UART_BUFSIZE, M_ZERO),
+               UART_BUFSIZE);
+  ringbuf_init(&state->tx_buf, kmalloc(M_DEV, UART_BUFSIZE, M_ZERO),
+               UART_BUFSIZE);
+
+  spin_init(&state->lock, 0);
+  cv_init(&state->rx_nonempty, "UART receive buffer not empty");
+  cv_init(&state->tx_nonfull, "UART transmit buffer not full");
+
+  resource_t *r = bus_alloc_resource(dev, RT_MEMORY, 0, UART0_BASE,
+                                     UART0_BASE + BCM2835_UART0_SIZE - 1,
+                                     BCM2835_UART0_SIZE, 0);
+
+  /* (pj) BCM2835_UART0_SIZE is much smaller than PAGESIZE */
+  bus_space_map(r->r_bus_tag, r->r_start, PAGESIZE, &r->r_bus_handle);
+
+  assert(r != NULL);
+
+  state->regs = r;
+
+  /* Disable UART0. */
+  bus_write_4(r, PL011COM_CR, 0);
+  /* Clear pending interrupts. */
+  bus_write_4(r, PL011COM_ICR, PL011_INT_ALLMASK);
+
+  /* TODO(pj) do magic with mail buffer */
+
+  gpio_function_select(r, 14, BCM2835_GPIO_ALT0);
+  gpio_function_select(r, 15, BCM2835_GPIO_ALT0);
+  gpio_set_pull(r, 14, BCM2838_GPIO_GPPUD_PULLOFF);
+  gpio_set_pull(r, 15, BCM2838_GPIO_GPPUD_PULLOFF);
+
+  /*
+   * Set integer & fractional part of baud rate.
+   * Divider = UART_CLOCK/(16 * Baud)
+   * Fraction part register = (Fractional part * 64) + 0.5
+   * UART_CLOCK = 3000000; Baud = 115200.
+   *
+   * Divider = 4000000/(16 * 115200) = 2.17 = ~2.
+   * Fractional part register = (.17 * 64) + 0.5 = 11.3 = ~11 = 0xb.
+   */
+
+  bus_write_4(r, PL011COM_IBRD, 2);
+  bus_write_4(r, PL011COM_FBRD, 0xb);
+
+  /* Enable FIFO & 8 bit data transmission (1 stop bit, no parity). */
+  bus_write_4(r, PL011COM_LCRH, PL01X_LCR_FEN | PL01X_LCR_8BITS);
+
+  /* Mask all interrupts. */
+  bus_write_4(r, PL011COM_IMSC, PL011_INT_ALLMASK);
+
+  /* Enable UART0, receive & transfer part of UART. */
+  bus_write_4(r, PL011COM_CR, PL01X_CR_UARTEN | PL011_CR_TXE | PL011_CR_RXE);
+
+  /* Enable receive interrupt. */
+  bus_write_4(r, PL011COM_IMSC, PL011_INT_RX);
+
+  state->irq = bus_alloc_irq(dev, 0, BCM2835_INT_UART0, RF_ACTIVE);
```

Comment:
I rewrote interrupt allocation & handling using fresh pieces of driver infrastructure. Please check again if the driver works.

Reply:
After some fixes.